File: src/main/java/xyz/devvydont/smprpg/items/blueprints/vanilla/ItemSpear.java
package xyz.devvydont.smprpg.items.blueprints.vanilla;

import org.bukkit.Material;
import org.bukkit.inventory.EquipmentSlotGroup;
import org.bukkit.inventory.ItemStack;
import xyz.devvydont.smprpg.attribute.AttributeWrapper;
import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.ItemClassification;
import xyz.devvydont.smprpg.items.attribute.AdditiveAttributeEntry;
import xyz.devvydont.smprpg.items.attribute.AttributeEntry;
import xyz.devvydont.smprpg.items.attribute.MultiplicativeAttributeEntry;
import xyz.devvydont.smprpg.items.base.VanillaAttributeItem;
import xyz.devvydont.smprpg.items.interfaces.IBreakableEquipment;
import xyz.devvydont.smprpg.services.ItemService;
import xyz.devvydont.smprpg.util.items.ToolGlobals;

import java.util.Collection;
import java.util.List;


public class ItemSpear extends VanillaAttributeItem implements IBreakableEquipment {

    public static double getSpearDamage(Material material) {
        return switch (material) {
            case NETHERITE_SPEAR -> 40;
            case DIAMOND_SPEAR -> 25;
            case GOLDEN_SPEAR -> 18;
            case IRON_SPEAR -> 15;
            case STONE_SPEAR, COPPER_SPEAR -> 10;
            case WOODEN_SPEAR -> 8;
            default -> 0;
        };
    }

    public static double getSpearDamage(CustomItemType itemType) {
        return switch (itemType) {
            case CustomItemType.TIN_SPEAR -> 9.0;
            case CustomItemType.SILVER_SPEAR -> 13.0;
            case CustomItemType.BRONZE_SPEAR -> 15.0;
            case CustomItemType.STEEL_SPEAR -> 18.0;
            case CustomItemType.ROSE_GOLD_SPEAR, CustomItemType.MITHRIL_SPEAR, CustomItemType.COBALT_SPEAR -> 20.0;
            case CustomItemType.TITANIUM_SPEAR, CustomItemType.TUNGSTEN_SPEAR -> 25.0;
            case CustomItemType.ADAMANTIUM_SPEAR -> 33.0;
            case CustomItemType.ORICHALCUM_SPEAR -> 35.0;
            case CustomItemType.DRAGONSTEEL_SPEAR -> 60.0;
            default -> 0;
        };
    }

    public static int getSpearRating(Material material) {
        return switch (material) {
            case NETHERITE_SPEAR -> ToolGlobals.NETHERITE_TOOL_POWER;
            case DIAMOND_SPEAR -> ToolGlobals.DIAMOND_TOOL_POWER;
            case GOLDEN_SPEAR -> ToolGlobals.GOLD_TOOL_POWER;
            case TRIDENT, IRON_SPEAR -> ToolGlobals.IRON_TOOL_POWER;
            case STONE_SPEAR -> ToolGlobals.STONE_TOOL_POWER;
            case WOODEN_SPEAR -> ToolGlobals.WOOD_TOOL_POWER;
            case COPPER_SPEAR -> ToolGlobals.COPPER_TOOL_POWER;
            default -> 1;
        };
    }

    public static double getSpearRecovery(Material material) {
        return switch (material) {
            case NETHERITE_SPEAR -> -0.85;
            case DIAMOND_SPEAR -> -0.8;
            case IRON_SPEAR, GOLDEN_SPEAR -> -0.75;
            case COPPER_SPEAR -> -0.7;
            case STONE_SPEAR -> -.65;
            case WOODEN_SPEAR -> -.6;
            default -> -1.0;
        };
    }

    public static double getSpearRecovery(CustomItemType itemType) {
        return switch (itemType) {
            case TIN_SPEAR -> -0.6;
            case SILVER_SPEAR -> -0.65;
            case BRONZE_SPEAR -> -0.7;
            case STEEL_SPEAR, MITHRIL_SPEAR, TITANIUM_SPEAR, ADAMANTIUM_SPEAR -> -0.75;
            case COBALT_SPEAR -> -0.75;
            case TUNGSTEN_SPEAR -> -0.8;
            case ORICHALCUM_SPEAR -> -0.85;
            case DRAGONSTEEL_SPEAR -> -0.9;
            default -> -1.0;
        };
    }

    public ItemSpear(ItemService itemService, Material material) {
        super(itemService, material);
    }

    @Override
    public ItemClassification getItemClassification() {
        return ItemClassification.SPEAR;
    }

    @Override
    public Collection<AttributeEntry> getAttributeModifiers(ItemStack item) {
        return List.of(
                new AdditiveAttributeEntry(AttributeWrapper.STRENGTH, getSpearDamage(material)),
                new MultiplicativeAttributeEntry(AttributeWrapper.ATTACK_SPEED, getSpearRecovery(material))
        );
    }

    @Override
    public int getPowerRating() {
        return getSpearRating(material);
    }

    @Override
    public EquipmentSlotGroup getActiveSlot() {
        return EquipmentSlotGroup.MAINHAND;
    }

    @Override
    public int getMaxDurability() {
        return switch (material) {
            case NETHERITE_SPEAR -> ToolGlobals.NETHERITE_TOOL_DURABILITY;
            case DIAMOND_SPEAR -> ToolGlobals.DIAMOND_TOOL_DURABILITY;
            case GOLDEN_SPEAR -> ToolGlobals.GOLD_TOOL_DURABILITY;
            case IRON_SPEAR -> ToolGlobals.IRON_TOOL_DURABILITY;
            case STONE_SPEAR -> ToolGlobals.STONE_TOOL_DURABILITY;
            case WOODEN_SPEAR -> ToolGlobals.WOOD_TOOL_DURABILITY;
            case COPPER_SPEAR -> ToolGlobals.COPPER_TOOL_DURABILITY;
            default -> 50_000;
        };
    }
}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/vanilla/ItemSpear.kt
package xyz.devvydont.smprpg.items.blueprints.vanilla

import org.bukkit.Material
import org.bukkit.inventory.EquipmentSlotGroup
import org.bukkit.inventory.ItemStack
import xyz.devvydont.smprpg.attribute.AttributeWrapper
import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.ItemClassification
import xyz.devvydont.smprpg.items.attribute.AdditiveAttributeEntry
import xyz.devvydont.smprpg.items.attribute.AttributeEntry
import xyz.devvydont.smprpg.items.attribute.MultiplicativeAttributeEntry
import xyz.devvydont.smprpg.items.base.VanillaAttributeItem
import xyz.devvydont.smprpg.items.interfaces.IBreakableEquipment
import xyz.devvydont.smprpg.items.interfaces.IRepairable
import xyz.devvydont.smprpg.items.interfaces.ISkillRequirement
import xyz.devvydont.smprpg.services.ItemService
import xyz.devvydont.smprpg.skills.SkillType
import xyz.devvydont.smprpg.util.items.ToolStats

class ItemSpear(itemService: ItemService, material: Material) : VanillaAttributeItem(itemService, material),
    IBreakableEquipment, IRepairable, ISkillRequirement {

    override val itemClassification: ItemClassification  get() = ItemClassification.SPEAR
    override val repairMaterial: MutableCollection<ItemStack>
        get() = when (material) {
            Material.NETHERITE_SPEAR -> mutableListOf(itemService.getCustomItem(Material.NETHERITE_INGOT))
            // Material.DIAMOND_SPEAR -> mutableListOf(itemService.getCustomItem(Material.DIAMOND))
            Material.GOLDEN_SPEAR -> mutableListOf(itemService.getCustomItem(Material.GOLD_INGOT))
            Material.IRON_SPEAR -> mutableListOf(itemService.getCustomItem(Material.IRON_INGOT))
            // Material.STONE_SPEAR -> mutableListOf(itemService.getCustomItem(Material.COBBLESTONE))
            Material.COPPER_SPEAR -> mutableListOf(itemService.getCustomItem(Material.COPPER_INGOT))
            Material.WOODEN_SPEAR -> mutableListOf(itemService.getCustomItem(Material.OAK_PLANKS), itemService.getCustomItem(Material.BIRCH_PLANKS), itemService.getCustomItem(Material.SPRUCE_PLANKS),
                itemService.getCustomItem(Material.JUNGLE_PLANKS), itemService.getCustomItem(Material.DARK_OAK_PLANKS), itemService.getCustomItem(Material.ACACIA_PLANKS),
                itemService.getCustomItem(Material.CRIMSON_PLANKS), itemService.getCustomItem(Material.WARPED_PLANKS), itemService.getCustomItem(Material.MANGROVE_PLANKS),
                itemService.getCustomItem(Material.CHERRY_PLANKS), itemService.getCustomItem(Material.PALE_OAK_PLANKS), itemService.getCustomItem(Material.BAMBOO_PLANKS))
            else -> mutableListOf()
        }
    override val skillRequirements: MutableMap<SkillType, Int> get() = when (material) {
        Material.NETHERITE_SPEAR -> mutableMapOf(Pair(SkillType.COMBAT, ToolStats.NETHERITE.skillReqLevel))
        Material.GOLDEN_SPEAR -> mutableMapOf(Pair(SkillType.COMBAT, ToolStats.GOLD.skillReqLevel))
        Material.IRON_SPEAR -> mutableMapOf(Pair(SkillType.COMBAT, ToolStats.IRON.skillReqLevel))
        Material.COPPER_SPEAR -> mutableMapOf(Pair(SkillType.COMBAT, ToolStats.COPPER.skillReqLevel))
        else -> mutableMapOf()
    }

    override fun getAttributeModifiers(item: ItemStack?): MutableCollection<AttributeEntry?> {
        return mutableListOf(
            AdditiveAttributeEntry(AttributeWrapper.STRENGTH, getSpearDamage(material)),
            MultiplicativeAttributeEntry(AttributeWrapper.ATTACK_SPEED, getSpearRecovery(material))
        )
    }

    override fun getPowerRating(): Int { return getSpearRating(material) }

    override fun getActiveSlot(): EquipmentSlotGroup { return EquipmentSlotGroup.MAINHAND }

    override fun getMaxDurability(): Int {
        return when (material) {
            Material.NETHERITE_SPEAR -> ToolStats.NETHERITE.durability
            // Material.DIAMOND_SPEAR -> ToolStats.DIAMOND.durability
            Material.GOLDEN_SPEAR -> ToolStats.GOLD.durability
            Material.IRON_SPEAR -> ToolStats.IRON.durability
            // Material.STONE_SPEAR -> ToolStats.STONE.durability
            Material.COPPER_SPEAR -> ToolStats.COPPER.durability
            Material.WOODEN_SPEAR -> ToolStats.WOOD.durability
            else -> 50000
        }
    }

    companion object {
        fun getSpearDamage(material: Material): Double {
            return when (material) {
                Material.NETHERITE_SPEAR -> 40
                Material.DIAMOND_SPEAR -> 25
                Material.GOLDEN_SPEAR -> 18
                Material.IRON_SPEAR -> 15
                Material.STONE_SPEAR, Material.COPPER_SPEAR -> 10
                Material.WOODEN_SPEAR -> 8
                else -> 0
            }.toDouble()
        }

        fun getSpearDamage(itemType: CustomItemType): Double {
            return when (itemType) {
                CustomItemType.TIN_SPEAR -> 9.0
                CustomItemType.SILVER_SPEAR -> 13.0
                CustomItemType.BRONZE_SPEAR -> 15.0
                CustomItemType.STEEL_SPEAR -> 18.0
                CustomItemType.ROSE_GOLD_SPEAR, CustomItemType.MITHRIL_SPEAR, CustomItemType.COBALT_SPEAR -> 20.0
                CustomItemType.TITANIUM_SPEAR, CustomItemType.TUNGSTEN_SPEAR -> 25.0
                CustomItemType.ADAMANTIUM_SPEAR -> 33.0
                CustomItemType.ORICHALCUM_SPEAR -> 35.0
                CustomItemType.DRAGONSTEEL_SPEAR -> 60.0
                else -> 0
            }.toDouble()
        }

        fun getSpearRating(material: Material): Int {
            return when (material) {
                Material.NETHERITE_SPEAR -> ToolStats.NETHERITE.power
                // Material.DIAMOND_SPEAR -> ToolStats.DIAMOND.power
                Material.GOLDEN_SPEAR -> ToolStats.GOLD.power
                Material.IRON_SPEAR -> ToolStats.IRON.power
                // Material.STONE_SPEAR -> ToolStats.STONE.power
                Material.COPPER_SPEAR -> ToolStats.COPPER.power
                Material.WOODEN_SPEAR -> ToolStats.WOOD.power
                else -> 1
            }
        }

        fun getSpearRecovery(material: Material): Double {
            return when (material) {
                Material.NETHERITE_SPEAR -> -0.85
                Material.DIAMOND_SPEAR -> -0.8
                Material.IRON_SPEAR, Material.GOLDEN_SPEAR -> -0.75
                Material.COPPER_SPEAR -> -0.7
                Material.STONE_SPEAR -> -.65
                Material.WOODEN_SPEAR -> -.6
                else -> -1.0
            }
        }

        fun getSpearRecovery(itemType: CustomItemType): Double {
            return when (itemType) {
                CustomItemType.TIN_SPEAR -> -0.6
                CustomItemType.SILVER_SPEAR -> -0.65
                CustomItemType.BRONZE_SPEAR -> -0.7
                CustomItemType.STEEL_SPEAR, CustomItemType.MITHRIL_SPEAR, CustomItemType.TITANIUM_SPEAR, CustomItemType.ADAMANTIUM_SPEAR -> -0.75
                CustomItemType.COBALT_SPEAR -> -0.75
                CustomItemType.TUNGSTEN_SPEAR -> -0.8
                CustomItemType.ORICHALCUM_SPEAR -> -0.85
                CustomItemType.DRAGONSTEEL_SPEAR -> -0.9
                else -> -1.0
            }
        }
    }
}
